src/pmdas/rocestat/pmdarocestat.python

```diff
@@ -0,0 +1,646 @@
+                nic_stats[key]['rx_bytes_bw'] = bw(cached_nic_stats[key]['rx_bytes'], nic_stats[key]['rx_bytes'])
+                nic_stats[key]['rx_pause_delta'] = per_sec(nic_stats[key]['rx_pause'], cached_nic_stats[key]['rx_pause'])
+        else:
+            for key in nic_stats.keys():
+                nic_stats[key]['tx_bytes_bw'] = 0
+                nic_stats[key]['rx_bytes_bw'] = 0
+                nic_stats[key]['rx_pause_delta'] = 0
+
+        for nic_lane, dict_stats in nic_stats.items():
+            nic = NicStats()
+            for k,v in dict_stats.items():
+                setattr(nic, k, v)
+
+            self.nic_stats[nic_lane] = nic
+
+    def refresh(self):
+        try:
+            self.refresh_mlnx_counters()
+            try:
+                self.refresh_nic_stats()
+            except:
+                self.log("Nic stats refresh failed")
+        except:
+            self.log("mlnx counters refresh failed")
+
+    def roce_stat_fetch(self):
+        ''' Called once per "fetch" PDU '''
+        self.refresh()
+        self.replace_indom(self.mlnx_indom, self.mlnx_counters)
+        self.replace_indom(self.nic_indom, self.nic_stats)
+
+    def roce_fetch_callback(self, cluster, item, inst):
+        """
+        Main fetch callback
+        """
+
+        if CLUSTER_HW <= cluster <= CLUSTER_HW_LINK:
+            return self.rocestat_fetch_mlnx_counters(item, inst)
+
+        if cluster == CLUSTER_NIC:
+            return self.rocestat_fetch_nic_stats(item, inst)
+
+        return [c_api.PM_ERR_PMID, 0]
+
+
+    def roce_store_callback(self, cluster, item, inst, val):
+        """
+        Store callback, executed when a request to write to a metric happens
+        """
+        self.log("storing values\n")
+
+
+    def __init__(self, name, domain, user, logfile):
+
+        super().__init__(name, domain, logfile)
+        if user is not None:
+            self.set_user(user)
+
+        self.connect_pmcd()
+
+        self.pmda_name = name
+
+        self.register_metrics(name)
+        self.set_fetch(self.roce_stat_fetch)
+
+        self.set_fetch_callback(self.roce_fetch_callback)
+        self.set_store_callback(self.roce_store_callback)
+
+    @staticmethod
+    def log(message):
+        PMDA.log(str(message))
+
+    @staticmethod
+    def err(message):
+        PMDA.err(message)
+
+    @classmethod
+    def in_pmda_setup(cls):
+        """Check if PMDA is in setup state"""
+        return os.environ.get('PCP_PYTHON_DOMAIN') or os.environ.get('PCP_PYTHON_PMNS')
+
+if __name__ == "__main__":
+    # This PMDA starts up in the "notready" state, see the Install script where
+    # the IPC protocol is ipc_prot="binary notready". See also pmcd(1) man page.
+    # The "binary notready" setting can also be manually configured in pmcd.conf.
+    pmda = ROCESTATPMDA(name='rocestat', domain=257, user="root", logfile='rocestat.log')
```

Comment:
Whats the reason for using "binary notready" in this PMDA?  Does it have a very slow startup?  (I can't see it, looks like it will be quick to me - in which case, its not needed)